man/unobb.1
.Dd August 5, 2018
.Dt UNOBB 1
.Os
.Sh NAME
.Nm unobb
.Nd Aspyr OBB virtual filesystem / archive extractor
.Sh SYNOPSIS
.Nm unobb
.Op Ar options
.Ar command
.Ar archive
.Op Ar
.Sh DESCRIPTION
.Nm
extracts Aspyr's OBB filesystem files, found in their Android
.Em Jade Empire
port, and OBB ZIP archives, found in Aspyr's Android
.Em Star Wars: Knights of the Old Republic
port.
.Pp
OBB files contain either a
.Dq virtual filesystem ,
with files split into chunks and compressed via zlib, or plain
ZIP archives.
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.El
.Bl -tag -width xx -compact
.It Ar command
.Bl -tag -width xx -compact
.It Cm l
List filesystem contents (without directories)
.It Cm v
List filesystem contents with full paths
.It Cm e
Extract files to current directory, stripping directories
.It Cm x
Extract files to current directory with full path, including directories.
.El
.It Ar archive
The OBB file to read.
.It Ar file
One or more files to extract.
If none are given, the whole archive is extracted.
.El
.Sh EXAMPLES
List all files contained in the filesystem
.Pa main.obb :
.Pp
.Dl $ unobb l main.obb
.Pp
Extract all files from the archive
.Pa main.obb :
.Pp
.Dl $ unobb e main.obb
.Pp
Extract a certain file from the archive
.Pa main.obb
with full path:
.Pp
.Dl $ unobb x main.obb a/certain/file.txt
.Sh SEE ALSO
.Xr unrim 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/unobb.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to extract Aspyr's OBB virtual filesystems / ZIP archives.
 */

#include <cstring>

#include <set>

#include "src/version/version.h"

#include "src/common/ustring.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/cli.h"
#include "src/common/readfile.h"
#include "src/common/scopedptr.h"

#include "src/aurora/obbfile.h"
#include "src/aurora/zipfile.h"

#include "src/archives/util.h"

#include "src/util.h"

enum Command {
	kCommandNone        = -1,
	kCommandList        =  0,
	kCommandListVerbose     ,
	kCommandExtract         ,
	kCommandExtractDir      ,
	kCommandMAX
};

const char *kCommandChar[kCommandMAX] = { "l", "v", "e", "x" };

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Command &command, Common::UString &archive, std::set<Common::UString> &files);

bool isPKZIP(Common::SeekableReadStream &stream);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		int returnValue = 1;
		Command command = kCommandNone;
		Common::UString archive;
		std::set<Common::UString> files;

		if (!parseCommandLine(args, returnValue, command, archive, files))
			return returnValue;

		Common::ScopedPtr<Common::SeekableReadStream> stream(new Common::ReadFile(archive));

		Common::ScopedPtr<Aurora::Archive> arc;
		if (isPKZIP(*stream))
			arc.reset(new Aurora::ZIPFile(stream.release()));
		else
			arc.reset(new Aurora::OBBFile(stream.release()));

		files = Archives::fixPathSeparator(files);

		if      (command == kCommandList)
			Archives::listFiles(*arc, Aurora::kGameIDUnknown, false);
		else if (command == kCommandListVerbose)
			Archives::listFiles(*arc, Aurora::kGameIDUnknown, true);
		else if (command == kCommandExtract)
			Archives::extractFiles(*arc, Aurora::kGameIDUnknown, false, files);
		else if (command == kCommandExtractDir)
			Archives::extractFiles(*arc, Aurora::kGameIDUnknown, true, files);

	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

namespace Common {
namespace CLI {
template<>
int ValGetter<Command &>::get(const std::vector<Common::UString> &args, int i, int) {
	_val = kCommandNone;
	for (int j = 0; j < kCommandMAX; j++) {
		if (!strcmp(args[i].c_str(), kCommandChar[j])) {
			_val = (Command) j;
			return 0;
		}
	}
	return -1;
}
}
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Command &command, Common::UString &archive, std::set<Common::UString> &files) {

	using Common::CLI::NoOption;
	using Common::CLI::Parser;
	using Common::CLI::ValGetter;
	using Common::CLI::makeEndArgs;

	NoOption cmdOpt(false, new ValGetter<Command &>(command, "command"));
	NoOption archiveOpt(false, new ValGetter<Common::UString &>(archive, "archive"));
	NoOption filesOpt(true, new ValGetter<std::set<Common::UString> &>(files, "files[...]"));
	Parser parser(argv[0], "Aspyr OBB virtual filesystem / archive extractor",
	              "Commands:\n"
	              "  l          List files (stripping directories)\n"
	              "  v          List files verbosely (with directories)\n"
	              "  e          Extract files to current directory, stripping directories\n"
	              "  x          Extract files to current directory, creating subdirectories\n",
	              returnValue,
	              makeEndArgs(&cmdOpt, &archiveOpt, &filesOpt));

	return parser.process(argv);
}

bool isPKZIP(Common::SeekableReadStream &stream) {
	const size_t pos = stream.pos();
	const bool pkzip = stream.readUint16BE() == MKTAG_16('P', 'K');

	stream.seek(pos);
	return pkzip;
}
